File: deck/base/code/native/node/runtime/mutex.ts
// Mutex runtime: an async flag lock. lock resolves immediately if free, else queues; unlock hands the lock to the next
// waiter or marks it free. The opaque handle a seed mutex holds is this object. Reached only through the public mutex API.
type SeedMutex = { locked: boolean; waiters: (() => void)[] }
const mutex = {
  make: (): SeedMutex => ({ locked: false, waiters: [] }),
  lock: (handle: SeedMutex): Promise<void> =>
    new Promise(ok => {
      if (!handle.locked) {
        handle.locked = true
        ok()
      } else handle.waiters.push(ok)
    }),
  unlock: (handle: SeedMutex): Promise<void> => {
    const waiter = handle.waiters.shift()
    if (waiter) waiter()
    else handle.locked = false
    return Promise.resolve()
  },
}

File: deck/base/code/native/rust/runtime/mutex.rs
// Mutex runtime: an atomic spinlock with cooperative yield, shared through an Arc (so the Clone-derived seed struct holds
// it, and lock / unlock are separate calls without a lifetime-bound guard). Reached only through the public mutex API.
mod mutex {
    use std::sync::atomic::{AtomicBool, Ordering};
    use std::sync::Arc;
    pub fn make() -> Arc<AtomicBool> {
        Arc::new(AtomicBool::new(false))
    }
    pub async fn lock(handle: Arc<AtomicBool>) {
        while handle
            .compare_exchange(false, true, Ordering::Acquire, Ordering::Relaxed)
            .is_err()
        {
            tokio::task::yield_now().await;
        }
    }
    pub async fn unlock(handle: Arc<AtomicBool>) {
        handle.store(false, Ordering::Release);
    }
}
